### PR TITLE
Project reference to package dependency code cleanup

### PR DESCRIPTION
**Pull request by chcosta**
This addresses issues brought up in a previously submitted PR and tracked with issue https://github.com/dotnet/buildtools/issues/830

- order (alphabetically) dependencies in project.json files and add the NuGet.Packaging package
- create a PackageItem class instead of using TaskItem for handling package information
- store package items as a dictionary
- error if package name clash (package information coming from multiple sources with differing versions)
- allow option to allow package name clash and take newest package version (UseNewestPackagesAvailable=true)
- when gathering information from a nupkg, get package information using the NuGet API's
- compare versions with NuGet API's
- fix regex in buildagainstpackages.targets

/cc @weshaggard @ericstj

### CODE REVIEW

**Comment by chcosta** [comment]
hmmm... I didn't intend for the "Port GetRuntimeTargets and GetRuntimeJsonValues for Package Testing" commit to be in this PR.  I'll look at removing those...

**Comment by chcosta** [comment]
I see that it's a no-op, I think it was just because I originally made the change unintentionally in my master branch and then cherry-picked it into another branch.

File: src/Microsoft.DotNet.Build.Tasks/project.json [hunk 1]
```diff
@@ -1,15 +1,16 @@
...
     "Microsoft.Build.Utilities.Core": "0.1.0-preview-00022",
     "Microsoft.Build.Tasks.Core": "0.1.0-preview-00022",
+    "NETStandard.Library": "1.5.0-rc2-24027",
     "Newtonsoft.Json": "7.0.1",
+    "NuGet.Packaging": "3.5.0-beta2-1456",
     "NuGet.Versioning": "3.5.0-beta2-1456",
     "NuProj": "0.10.4-beta-gf7fc34e7d8",
-    "NETStandard.Library": "1.5.0-rc2-24027"
+    "System.Reflection.Metadata": "1.3.0-rc3-24128-00",
+    "System.Dynamic.Runtime": "4.0.11-rc3-24128-00",
```
Comment (chcosta) [inline src/Microsoft.DotNet.Build.Tasks/project.json]: Haha,  I only touched this file to add nuget.packaging and to put these entries into alphabetical order but it appears I didn't get the alphabetical order bit entirely correct.  I'll submit a tweak for this tomorrow.

**Comment by chcosta** [comment]
@ericstj, can you take a look at this PR?

File: src/Microsoft.DotNet.Build.Tasks/AddDependenciesToProjectJson.cs [hunk 2]
```diff
@@ -165,35 +212,36 @@ private IEnumerable<ITaskItem> GatherPackageInformationFromVersionsFile(string v
...
+
+                return new PackageItem(m.Groups["name"].Value, nuGetVersion);
             }
-            return packageItem;
+            return null;
+        }
+
+        private PackageItem CreatePackageItemFromNupkg(string package)
+        {
+            PackageArchiveReader archiveReader = new PackageArchiveReader(package);
```
Comment (ericstj) [inline src/Microsoft.DotNet.Build.Tasks/AddDependenciesToProjectJson.cs]: Need a `using` here.

File: src/Microsoft.DotNet.Build.Tasks/AddDependenciesToProjectJson.cs [hunk 3]
```diff
@@ -64,14 +75,16 @@ public override bool Execute()
...
             _packageNameRegex = new Regex(PackageNameRegex);
 
+            VersionComparer comparer = new VersionComparer(VersionComparison.VersionRelease);
+
             // Retrieve package information from a package drop location
             if (PackagesDrops != null &&
                 PackagesDrops.Length > 0)
             {
-                packageInformation.AddRange(GatherPackageInformationFromDrops(PackagesDrops));
+                AddPackageItemsToDictionary(ref packageInformation, GatherPackageInformationFromDrops(PackagesDrops, comparer), comparer);
```
Comment (ericstj) [inline src/Microsoft.DotNet.Build.Tasks/AddDependenciesToProjectJson.cs]: You don't need `ref`

File: src/Microsoft.DotNet.Build.Tasks/AddDependenciesToProjectJson.cs [hunk 4]
```diff
@@ -64,14 +75,16 @@ public override bool Execute()
                 return false;
             }
 
-            List<ITaskItem> packageInformation = new List<ITaskItem>();
+            Dictionary<string, PackageItem> packageInformation = new Dictionary<string, PackageItem>();
             _packageNameRegex = new Regex(PackageNameRegex);
 
+            VersionComparer comparer = new VersionComparer(VersionComparison.VersionRelease);
```
Comment (ericstj) [inline src/Microsoft.DotNet.Build.Tasks/AddDependenciesToProjectJson.cs]: You only ever set this to `VersionComparer(VersionComparison.VersionRelease)`.  Can you make it an instance field and avoid passing it around everywhere?

File: src/Microsoft.DotNet.Build.Tasks/AddDependenciesToProjectJson.cs [hunk 5]
```diff
@@ -165,35 +212,36 @@ private IEnumerable<ITaskItem> GatherPackageInformationFromVersionsFile(string v
...
+                    PackageItem packageItem = CreatePackageItemFromString(packageVersionTokens[0], packageVersionTokens[1]);
+                    AddPackageItemToDictionary(ref packageItems, packageItem, comparer);
                 }
             }
-            return packageNameItems;
+            return packageItems;
         }
 
-        private TaskItem CreatePackageItemFromString(string package)
+        private PackageItem CreatePackageItemFromString(string package)
```
Comment (ericstj) [inline src/Microsoft.DotNet.Build.Tasks/AddDependenciesToProjectJson.cs]: Delete this method.

File: src/Microsoft.DotNet.Build.Tasks/AddDependenciesToProjectJson.cs [hunk 6]
```diff
@@ -165,35 +212,36 @@ private IEnumerable<ITaskItem> GatherPackageInformationFromVersionsFile(string v
...
+
+        private PackageItem CreatePackageItemFromNupkg(string package)
+        {
+            PackageArchiveReader archiveReader = new PackageArchiveReader(package);
+            PackageIdentity identity = archiveReader.GetIdentity();
+            return new PackageItem(identity.Id, identity.Version);
         }
 
-        private TaskItem CreatePackageItemFromString(string id, string version)
+        private PackageItem CreatePackageItemFromString(string id, string version)
```
Comment (ericstj) [inline src/Microsoft.DotNet.Build.Tasks/AddDependenciesToProjectJson.cs]: Consider renaming this.

**Comment by ericstj** [comment]
LGTM

**Pull request merged**